### ISSUE TITLE
[Bug] Multiprocessing pool hangs when Beartype exception is raised

### ISSUE THREAD

**Issue by avolchek** [beartype/beartype#382]
Hi!
When code below is executed exception somewhere inside `multiprocessing` is raised and it hangs:
```
from beartype.door import die_if_unbearable
import multiprocessing


def foo(x):
  die_if_unbearable(x, str)


if __name__ == '__main__':
  with multiprocessing.Pool(processes=1) as pool:
    list(pool.imap(foo, [42]))
```

Traceback:
```
Exception in thread Thread-3 (_handle_results):
Traceback (most recent call last):
  File "/Users/avolchek/.pyenv/versions/3.10.13/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
    self.run()
  File "/Users/avolchek/.pyenv/versions/3.10.13/lib/python3.10/threading.py", line 953, in run
    self._target(*self._args, **self._kwargs)
  File "/Users/avolchek/.pyenv/versions/3.10.13/lib/python3.10/multiprocessing/pool.py", line 579, in _handle_results
    task = get()
  File "/Users/avolchek/.pyenv/versions/3.10.13/lib/python3.10/multiprocessing/connection.py", line 251, in recv
    return _ForkingPickler.loads(buf.getbuffer())
TypeError: BeartypeCallHintViolation.__init__() missing 1 required positional argument: 'culprits'
```

I think that it may be related to https://stackoverflow.com/a/28335286

**Comment by avolchek** [comment]
it seems that the following patch fixes hanging
```
--- a/Users/avolchek/.pyenv/versions/3.10.13/lib/python3.10/site-packages/beartype/roar/_roarexc.py
+++ a/Users/avolchek/.pyenv/versions/3.10.13/lib/python3.10/site-packages/beartype/roar/_roarexc.py
@@ -49,7 +49,7 @@ class BeartypeException(Exception, metaclass=_ABCMeta):
     '''

     # ..................{ INITIALIZERS                       }..................
-    def __init__(self, message: str) -> None:
+    def __init__(self, message: str, *args, **kwargs) -> None:
         '''
         Initialize this exception.

@@ -88,7 +88,7 @@ class BeartypeException(Exception, metaclass=_ABCMeta):
             message = f'{message[0].upper()}{message[1:]}'

         # Defer to the superclass constructor.
-        super().__init__(message)
+        super().__init__(message, *args, **kwargs)

         # Sanitize the fully-qualified module name of the class of this
         # exception. See the docstring for justification.
@@ -585,7 +585,7 @@ class BeartypeCallHintViolation(BeartypeCallHintException):
         from beartype._util.py.utilpyweakref import make_obj_weakref_and_repr

         # Initialize the superclass with the passed message.
-        super().__init__(message)
+        super().__init__(message, culprits)

         #FIXME: Unit test us up, please.
         # If the culprits are *NOT* a tuple, raise an exception.
```

but smth goes wrong with formatting in terminal:

```

multiprocessing.pool.RemoteTraceback:
"""
Traceback (most recent call last):
  File "/Users/avolchek/.pyenv/versions/3.10.13/lib/python3.10/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/Users/avolchek/beartype_bug/kek.py", line 9, in foo
    die_if_unbearable(x, str)
  File "/Users/avolchek/.pyenv/versions/3.10.13/lib/python3.10/site-packages/beartype/door/_doorcheck.py", line 106, in die_if_unbearable
    func_raiser(obj)  # pyright: ignore[reportUnboundVariable]
  File "<@beartype(__beartype_checker_0) at 0x1047168f0>", line 19, in __beartype_checker_0
beartype.roar.BeartypeDoorHintViolation: ("Die_if_unbearable() value \x1b[1m\x1b[31m42\x1b[0m violates type hint \x1b[1m\x1b[32m<class 'str'>\x1b[0m, as \x1b[1m\x1b[33mint \x1b[0m\x1b[1m\x1b[31m42\x1b[0m not instance of \x1b[1m\x1b[32mstr\x1b[0m.", (42,))
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/avolchek/beartype_bug/kek.py", line 14, in <module>
    list(pool.imap(foo, [42]))
  File "/Users/avolchek/.pyenv/versions/3.10.13/lib/python3.10/multiprocessing/pool.py", line 873, in next
    raise value
beartype.roar.BeartypeDoorHintViolation: ("Die_if_unbearable() value \x1b[1m\x1b[31m42\x1b[0m violates type hint \x1b[1m\x1b[32m<class 'str'>\x1b[0m, as \x1b[1m\x1b[33mint \x1b[0m\x1b[1m\x1b[31m42\x1b[0m not instance of \x1b[1m\x1b[32mstr\x1b[0m.", (42,))
```

**Comment by leycec** [comment]
**Holy balls of @beartype.** Thanks a heap overflow for this tremendous volunteerism, @avolchek. Without your phenomenal changeset, this would've been *super* non-trivial for me to resolve. You saved my weekend. I'm indebted and will buy you pancakes stuffed with maple syrup when you visit Canada. As it is, I'll do my utmost to:

* Immediately resolve this.
* Release this resolution (*...probably early next week*) with our upcoming @beartype 0.19.0 release.

And thanks for noting the oddball exception message, too. It looks like the original "bad" input argument `[42]` passed to `pool.imap()` is being coerced somewhere along the way into a tuple `(42,)` (*...presumably, as the @beartype-specific `culprit` exception argument), which is then being unpickled into yet another 2-tuple.

Strange stuff. But that's `multiprocessing` for you, huh? I still can't believe they're using the broken `pickle` module rather than the third-party `dill` package after all these years. My emoji weeps bitter tears that are salty and manly for Python! :sob:

**Comment by leycec** [comment]
**Resolved by 11be32d9c50a.** Welp, that was madness. You do *not* want to know how we resolved the unreadable exception message issue. Very well! It took me four hours to unparse this – but here's what I eventually settled on:

```python
class BeartypeException(Exception, metaclass=_ABCMeta):
    ...

    def __str__(self) -> str:
        '''
        Human-readable message describing this exception.

        Note that this dunder method *should* be redundant. Since the
        :meth:`__init__` method of this subclass explicitly passes this same
        exact message to the :meth:`__init__` method of the :exc:`Exception`
        superclass, there should be *no* need to expose this message yet again
        by defining this dunder method.

        Indeed, this dunder method *is* redundant for almost all edge cases --
        except one. For unknown reasons, the standard :mod:`multiprocessing`
        package renders a non-human-readable tuple rather than this
        human-readable message when a Python subprocess forked by a
        multiprocessing pool raises an instance of this exception resembling:

            beartype.roar.BeartypeDoorHintViolation: ('Die_if_unbearable() value
            \x1b[1m\x1b[31m42\x1b[0m violates type hint
            \x1b[1m\x1b[32mtyping.List[str]\x1b[0m, as \x1b[1m\x1b[33mint
            \x1b[0m\x1b[1m\x1b[31m42\x1b[0m not instance of
            \x1b[1m\x1b[32mlist\x1b[0m.', (42,))

        Clearly, this has something inexplicable to do with exception pickling
        internally performed by the :mod:`multiprocessing` package. Just as
        clearly, we have neither the inclination nor the patience to get to the
        bottom of this. Indeed, this implementation is inspired by the standard
        :exc:`subprocess.CalledProcessError` exception subclass -- which defines
        the ``__str__()`` dunder method in a similar manner and (presumably) for
        the exact same reasons. This is a mess that we want nothing to do with.
        '''

        # Return the first parameter passed to the superclass __init__() method,
        # guaranteed to be the desired human-readable exception message.
        return self.args[0]
```

Only Lieutenant Commander Spock knows what happened here.

![](https://media1.tenor.com/m/vsq59JdcqHwAAAAd/star-trek-spock.gif)
